lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs

```diff
@@ -2387,3 +2440,388 @@ outputsCreatedByTx tx
+  where
+    -- The 'currentTip' field is not currently used by 'applyBlocks'.
+    --
+    currentTip :: BlockHeader
+    currentTip = shouldNotEvaluate "currentTip"
+
+    shouldNotEvaluate :: String -> a
+    shouldNotEvaluate name = error $ unwords
+        [name, "was unexpectedly evaluated"]
+
+-- | Extracts a list of filtered transactions from the result of 'applyBlocks'.
+--
+-- The returned transactions are listed in the same order that they appear
+-- within the filtered blocks.
+--
+applyBlocksFilteredTxs
+    :: NonEmpty ([FilteredBlock], (DeltaWallet s, Wallet s)) -> [Tx]
+applyBlocksFilteredTxs = mconcat . mconcat . NE.toList
+    . fmap (fmap (fmap fst . view #transactions) . fst)
+
+-- | Extracts out the last UTxO set returned by 'applyBlocks'.
+--
+-- This UTxO set represents the final state of the wallet after applying
+-- just those transactions that are relevant to the wallet.
+--
+applyBlocksLastUTxO
+    :: NonEmpty ([FilteredBlock], (DeltaWallet s, Wallet s)) -> UTxO
+applyBlocksLastUTxO = utxo . snd . snd . NE.last
+
+-- | Removes a fee from a transaction.
+--
+-- Useful in situations where we want to compare transactions without comparing
+-- their fees.
+--
+nullifyFee :: Tx -> Tx
+nullifyFee = set #fee Nothing
+
+--------------------------------------------------------------------------------
+-- Convenience functions for manipulating UTxO sets
+--------------------------------------------------------------------------------
+
+-- | Filters a UTxO set for entries that are expected to be relevant to a
+--   particular wallet.
+--
+ourUTxO :: forall s. (IsOurs s Address) => s -> UTxO -> UTxO
+ourUTxO s u = evalState (UTxO.filterByAddressM isOurAddressM u) s
+  where
+    isOurAddressM :: Address -> State s Bool
+    isOurAddressM = fmap isJust . state . isOurs
+
+--------------------------------------------------------------------------------
+-- Verifying transactions returned by 'applyBlocks'
+--------------------------------------------------------------------------------
+
+--------------------------------------------------------------------------------
+-- The following properties verify that 'applyBlocks':
+--
+-- - returns all transactions that are relevant to the wallet, and no others.
+-- - returns transactions in the same order that they appear on the blockchain.
+--------------------------------------------------------------------------------
+
+-- Scenario: some transactions within the block sequence are expected to be
+-- relevant to the wallet, but not all.
+--
+prop_applyBlocks_filteredTxs_someOurs
+    :: Pretty (IsOursIf2 Address RewardAccount)
+    -> Pretty BlockSeq
+    -> Property
+prop_applyBlocks_filteredTxs_someOurs (Pretty someOurs) (Pretty blockSeq)
+    -- TODO: we currently ignore transaction fees, as under some circumstances
+    -- the 'applyBlocks' function can modify fee values:
+    = fmap nullifyFee ourTxsReturned ====
+      fmap nullifyFee ourTxsExpected
```

Comment:
Do we really want do deep-compare transactions or maybe just comparing their ids (hashes) suffice?

Reply:
> Do we really want do deep-compare transactions or maybe just comparing their ids (hashes) suffice?

Given the current type of `applyBlocks`, I think it's useful to verify that it doesn't modify transactions in any meaningful way.

If `applyBlocks` were modified to return a set of transaction identifiers (i.e., a subset of the blocks supplied), or if we had a type-level proof that the function was incapable of modifying transactions, then we could drop this comparison entirely.